sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/V1Writes.scala

```diff
@@ -0,0 +1,148 @@
+      dataColumns: Seq[Attribute],
+      options: Map[String, String]): Option[WriterBucketSpec] = {
+    bucketSpec.map { spec =>
+      val bucketColumns = spec.bucketColumnNames.map(c => dataColumns.find(_.name == c).get)
+      if (options.getOrElse(BucketingUtils.optionForHiveCompatibleBucketWrite, "false") ==
+        "true") {
+        // Hive bucketed table: use `HiveHash` and bitwise-and as bucket id expression.
+        // Without the extra bitwise-and operation, we can get wrong bucket id when hash value of
+        // columns is negative. See Hive implementation in
+        // `org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorUtils#getBucketNumber()`.
+        val hashId = BitwiseAnd(HiveHash(bucketColumns), Literal(Int.MaxValue))
+        val bucketIdExpression = Pmod(hashId, Literal(spec.numBuckets))
+
+        // The bucket file name prefix is following Hive, Presto and Trino conversion, so this
+        // makes sure Hive bucketed table written by Spark, can be read by other SQL engines.
+        //
+        // Hive: `org.apache.hadoop.hive.ql.exec.Utilities#getBucketIdFromFile()`.
+        // Trino: `io.trino.plugin.hive.BackgroundHiveSplitLoader#BUCKET_PATTERNS`.
+        val fileNamePrefix = (bucketId: Int) => f"$bucketId%05d_0_"
+        WriterBucketSpec(bucketIdExpression, fileNamePrefix)
+      } else {
+        // Spark bucketed table: use `HashPartitioning.partitionIdExpression` as bucket id
+        // expression, so that we can guarantee the data distribution is same between shuffle and
+        // bucketed data source, which enables us to only shuffle one side when join a bucketed
+        // table and a normal one.
+        val bucketIdExpression = HashPartitioning(bucketColumns, spec.numBuckets)
+          .partitionIdExpression
+        WriterBucketSpec(bucketIdExpression, (_: Int) => "")
+      }
+    }
+  }
+
+  def getBucketSortColumns(
+      bucketSpec: Option[BucketSpec], dataColumns: Seq[Attribute]): Seq[Attribute] = {
+    bucketSpec.toSeq.flatMap {
+      spec => spec.sortColumnNames.map(c => dataColumns.find(_.name == c).get)
+    }
+  }
+
+  def getSortOrder(
+      outputColumns: Seq[Attribute],
+      partitionColumns: Seq[Attribute],
+      numStaticPartitions: Int,
+      bucketSpec: Option[BucketSpec],
+      options: Map[String, String]): Seq[SortOrder] = {
+    val partitionSet = AttributeSet(partitionColumns)
+    val dataColumns = outputColumns.filterNot(partitionSet.contains)
+    val writerBucketSpec = getBucketSpec(bucketSpec, dataColumns, options)
+    val sortColumns = getBucketSortColumns(bucketSpec, dataColumns)
+
+    assert(partitionColumns.size >= numStaticPartitions)
+    // We should first sort by partition columns, then bucket id, and finally sorting columns.
+    (partitionColumns.takeRight(partitionColumns.size - numStaticPartitions) ++
+      writerBucketSpec.map(_.bucketIdExpression) ++ sortColumns)
+      .map(SortOrder(_, Ascending))
+  }
+
+  def prepareQuery(
```

Comment:
Actually after taking a deeper look, `LogicalPlan.outputOrdering` was introduced to eliminate unnecessary sort in logical planning (https://github.com/apache/spark/pull/20560), and only several operators preserves ordering (like `Filter`, `Project`), so it won't cause correctness issue here.

But the problem of `LogicalPlan.outputOrdering` is it being too conservative. We may add unnecessary sort here for complex queries (e.g. query with sort merge join on partition columns before writing to table with dynamic partitions)

Reply:
thank you @c21 for pointing out this and I see what you concern about. The reason I used the `LogicalPlan.outputOrdering` is:
- Add sort at logical side has benefits if the plan exists a sort. e.g.
  ```
  InsertIntoTable (partition) 
    Sort (not dynamic columns)
      ....
  ```
  We can eliminate the user specified sort using `EliminateSorts` in `Optimizer`. But if we add the sort at physical plan, we will do the sort twice even the first sort has no effect.
- For now, I prefer to keep the same approach with `V2Writes` which also add the required ordering even distribution at logical side. We can optimize them together if we find a more better approach in future.
- I thnk it's safe and no perf regression that add a sort at logical side. Since we have the `RemoveRedundantSorts` at physical side, that rule can remove the sort we added if it's uncessary (e.g. sort + smj with dynamic partitions).